lib/libxdp/libxdp.c

```diff
@@ -2157,11 +2235,11 @@ int xdp_multiprog__detach(struct xdp_multiprog *mp)
 	if (!mp || !mp->is_loaded)
 		return -EINVAL;
 
-	err = xdp_multiprog__attach(mp, NULL, mp->attach_mode);
+	err = xdp_multiprog__attach(mp, NULL, XDP_MODE_UNSPEC);
 	if (err)
 		return err;
 
-	if (!mp->is_legacy)
+	if (!mp->is_legacy && mp->main_prog)
```

Comment:
How can main_prog be unset if is_legacy is false?

Reply:
Oh, if there's *only* an HW prog. In that case I think we should also set is_legacy...

Reply:
I tried to keep the is_legacy so that it corresponds to the main_prog only, since hw_prog is always "legacy" and there might be some ambiguity if both main_prog and hw_prog would be attached.

Reply:
Hmm, it does become a bit weird that we can basically have a 'null' value for main_prog now with hw-mode only. But I think is_legacy should mean "doesn't have a dispatcher". Which can then either mean "doesn't have a dispatcher because it's a single program attached in software" in which case main_prog is != null, or it can mean "doesn't have a dispatcher because it's only an hw prog", in which case main_prog == null.

In any case, these semantics should be documented; please add that to both the commit message and the README file :)

Reply:
Also, is is_legacy field really needed? Isn't is_legacy equivalent to first_prog == NULL?

Reply:
Yeah, guess so; if you want to get rid of it, go ahead, but you'd need to change all the internal references to use xdp_multiprog__is_legacy() instead first :)